compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -265,32 +125,61 @@ pub fn is_const_evaluatable<'tcx>(
     }
 }
 
-#[instrument(skip(tcx), level = "debug")]
+#[instrument(skip(infcx, tcx), level = "debug")]
 fn satisfied_from_param_env<'tcx>(
     tcx: TyCtxt<'tcx>,
-    ct: AbstractConst<'tcx>,
+    infcx: &InferCtxt<'tcx>,
+    ct: ty::Const<'tcx>,
     param_env: ty::ParamEnv<'tcx>,
 ) -> Result<bool, NotConstEvaluatable> {
+    // Try to unify with each subtree in the AbstractConst to allow for
+    // `N + 1` being const evaluatable even if theres only a `ConstEvaluatable`
+    // predicate for `(N + 1) * 2`
+    struct Visitor<'a, 'tcx> {
+        ct: ty::Const<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+
+        infcx: &'a InferCtxt<'tcx>,
+    }
+    impl<'a, 'tcx> TypeVisitor<'tcx> for Visitor<'a, 'tcx> {
+        type BreakTy = ();
+        fn visit_const(&mut self, c: ty::Const<'tcx>) -> ControlFlow<Self::BreakTy> {
+            if c.ty() == self.ct.ty()
+                && let Ok(_nested_obligations) = self
+                    .infcx
+                    .at(&ObligationCause::dummy(), self.param_env)
+                    .eq(c, self.ct)
+            {
+                //let obligations = nested_obligations.into_obligations();
+                ControlFlow::BREAK
+            } else if let ty::ConstKind::Expr(e) = c.kind() {
+                e.visit_with(self)
+            } else {
+                ControlFlow::CONTINUE
+            }
+        }
+    }
+
     for pred in param_env.caller_bounds() {
         match pred.kind().skip_binder() {
-            ty::PredicateKind::ConstEvaluatable(uv) => {
-                if let Some(b_ct) = AbstractConst::from_const(tcx, uv)? {
-                    let const_unify_ctxt = ConstUnifyCtxt { tcx, param_env };
+            ty::PredicateKind::ConstEvaluatable(ce) => {
+                let ty::ConstKind::Unevaluated(_) = ce.kind() else {
+                    continue
+                };
+                let b_ct = tcx.expand_abstract_consts2(ce);
+                if !matches!(b_ct.kind(), ty::ConstKind::Expr(_)) {
+                    continue;
+                } else if b_ct.references_error() {
+                    let guard = tcx.sess.delay_span_bug(DUMMY_SP, "ct references error");
+                    return Err(NotConstEvaluatable::Error(guard));
+                };
```

Comment:
```suggestion
                b_ct.error_reported().map_err(NotConstEvaluatable::Error)?;
```

Reply:
that ought to fix most of the test failures, a `ConstKind::Unevaluated` cant always be normalized to an expr i.e. `<T as Trait>::ASSOC` is "as normalized as it can be" unless we can resolve the specific `T: Trait` impl, and even then we currently dont support building abstract consts out of the actual value of an assoc const

Reply:
Will try this later tonight,

slowly merging all of lcnr's changes in